pytorch_lightning/strategies/sharded.py

```diff
@@ -40,6 +40,23 @@ class DDPShardedStrategy(DDPStrategy):
     distributed_backend = _StrategyType.DDP_SHARDED
     _REDUCE_BUFFER_SIZE_DEFAULT: int = 2 ** 23  # 8M
 
+    def setup(self, trainer: "pl.Trainer") -> None:
+        # share ddp pids to all processes
+        self._rank_0_has_called_call_children_scripts = self.broadcast(self._rank_0_has_called_call_children_scripts)
+        if self._should_run_deadlock_detection():
+            self._share_information_to_prevent_deadlock()
+
+        # move the model to the correct device
+        self.model_to_device()
+
+        if self.sync_batchnorm:
+            self.model = self.configure_sync_batchnorm(self.model)
+
+        # skip wrapping the model if we are not fitting as no gradients need to be exchanged
+        trainer_fn = trainer.state.fn
+        if trainer_fn == TrainerFn.FITTING:
+            self.configure_ddp()
```

Comment:
Sharded inheriting from DDP makes this unnecessarily complicated considering they have different behaviors. This is why issues like https://github.com/PyTorchLightning/pytorch-lightning/issues/11863 are important

Reply:
We are missing a call to super().setup() here. Which means we won't setup optimizers there which is fine I guess because we do that in configure_ddp for this strategy. 

Just want to make sure you are aware that the missing super.setup()  means we won't call `accelerator.setup()` and `self.setup_precision_plugin() `

Reply:
@awaelchli - I am holding off on this until https://github.com/PyTorchLightning/pytorch-lightning/pull/11985 is reviewed & merged. then i'll redo the optimizer initialization, precision plugin setup, and accelerator setup

Reply:
I cannot call `super().setup()` because that will all DDPStrategy.setup() which does something completely different: it wraps the model in DistributedDataParallel, not ShardedDataParallel!

Reply:
@awaelchli I don't even think we can rely on `super().setup()` inside of DDP since we should call the accelerator.setup(trainer) *before* we call `model_to_device` to handle any accelerator validation. 
https://github.com/PyTorchLightning/pytorch-lightning/blob/b5fe056765ce71b237c641a1b159c346e5e8bd33/pytorch_lightning/accelerators/gpu.py#L46-L50

at this point i really dislike how `setup` is packaged on the base strategy. i'd much prefer calling the individual methods at different points rather than try to dance around them.